module/mailnews/newsread.c

```diff
@@ -394,3 +412,100 @@ Bool DateFromSeconds(long seconds, char *str)
 	   t->tm_mday, t->tm_year + 1900, t->tm_hour, t->tm_min);
    return True;
 }
+
+void OnColumnClick(LPNMLISTVIEW pLVInfo)
+{
+    int nSortColumn = pLVInfo->iSubItem;
+    BOOL bSortAscending = TRUE;
+
+    if (nSortColumn == currentSortColumn) {
+        // Toggle sort direction if the same column is clicked again
+        currentSortAscending = !currentSortAscending;
+    } else {
+        // Set ascending order for new column
+        currentSortColumn = nSortColumn;
+        currentSortAscending = TRUE;
+    }
+
+    LPARAM lParamSort = (currentSortAscending ? 1 : -1) * (currentSortColumn + 1);
+
+    // Update header text
+    UpdateColumnHeaders(pLVInfo->hdr.hwndFrom, currentSortColumn, currentSortAscending);
+
+    // Sort list
+    ListView_SortItems(pLVInfo->hdr.hwndFrom, CompareListItems, lParamSort);
+}
+
+int CALLBACK CompareListItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
+{
+   BOOL bSortAscending = (lParamSort > 0);
+   int nColumn = abs(lParamSort) - 1;
+
+   NewsArticle *article1 = (NewsArticle *)lParam1;
+   NewsArticle *article2 = (NewsArticle *)lParam2;
+
+   // `ListView_SortItems` is unstable, so we append the article index for stability
+   char title1[MAX_SUBJECT + 10];
+   sprintf(title1, "%s %d", article1->title, article1->num);
+   char title2[MAX_SUBJECT + 10];
+   sprintf(title2, "%s %d", article2->title, article2->num);
+
+   char poster1[MAXUSERNAME + 10];
+   sprintf(poster1, "%s %d", article1->poster, article1->num);
+   char poster2[MAXUSERNAME + 10];
+   sprintf(poster2, "%s %d", article2->poster, article2->num);
+
+   switch (nColumn)
+   {
+   case COL_TITLE:
+      if (bSortAscending)
+         return stricmp(title1, title2);
+      return stricmp(title2, title1);
+   case COL_POSTER:
+      if (bSortAscending)
+         return stricmp(poster1, poster2);
+      return stricmp(poster2, poster1);
+   case COL_TIME:
+      if (bSortAscending)
+         return article1->time - article2->time;
+      return article2->time - article1->time;
+   }
+
+   return 0;
+}
+
+void UpdateColumnHeaders(HWND hListView, int sortedColumn, BOOL sortAscending)
+{
+    HWND hHeader = ListView_GetHeader(hListView);
+    int columnCount = Header_GetItemCount(hHeader);
+
+    for (int i = 0; i < columnCount; i++)
+    {
+        TCHAR text[256];
```

Comment:
TCHAR is a hideous Windows-ism.  Use char if the later code will still work with it.

Where did 256 come from?  Use a named constant.